Dashboard: Replace datasource name with a reference object

ryantxu opened the pull request:
Fixes https://github.com/grafana/grafana/issues/1664, See also https://github.com/grafana/grafana/issues/32853

We currently store the datasource *name* in the dashboard, this has a few bad effects:
1. not enough information in a dashboard to know if the queries are valid without a database lookup.  See https://github.com/grafana/grafana/issues/33761
2. renaming a datasource will break all references

This replaces our existing string object with a datasourceRef:
```
export interface DatasourceRef {
  type?: string;
  uid?: string;
}
```
Although we *could* simply replace our existing name with the uid, using an object with type makes defining dashboards programatically MUCH more natural.  Specifically, this allows us to day "I need a loki datasource, but I don't care which one".  Also the type parameter will allow validating the query syntax without without needing to lookup `uid` > `datasource.pluginType`


TODO:

- [X] migrate variable queries
- [x] dashboard queries... 
- [x] Unified alerting seems to be working and tests pass (after a code update by torkel, se changes in expr/graph and expr/nodes)
- [x] Updated legacy alerting alert rule extractor 
- [ ] What should dashboard exporter do? 
- [ ] Annotation queries

ryantxu committed:
use datasource object

ryantxu committed:
use datasource object

ryantxu committed:
migrate variables

ryantxu committed:
more futzing

ryantxu committed:
update snapshot

ryantxu commented on packages/grafana-data/src/types/query.ts:
```diff
@@ -0,0 +1,49 @@
+   *
+   * RUNTIME!!! (should not be saved in JSON!!!!)
+   */
+  dataTopic?: DataTopic;
```
@torkelo -- should this ever be stored in the dashboard?  or is it a runtime property?

ryantxu committed:
use expression uid

ryantxu committed:
update snapshot

ryantxu committed:
remove dataTopic

torkelo commented on packages/grafana-data/src/types/datasource.ts:
```diff
@@ -204,6 +205,9 @@ abstract class DataSourceApi<
     this.type = instanceSettings.type;
     this.meta = {} as DataSourcePluginMeta;
     this.uid = instanceSettings.uid;
+    if (!this.uid) {
+      this.uid = this.name;
+    }
```
what is this condition for? Some of the built in (-- Dashboard --, -- Grafana --) data sources?

ryantxu replied:
yes the builtin ones do not have uid (or id)

torkelo reviewed (commented):
This looks very promising, awesome job

ryantxu committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

dependabot[bot] committed:
Chore(deps): Bump handlebars from 4.7.6 to 4.7.7 (#33819)

ryantxu committed:
panel query runner test

ryantxu committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

ryantxu committed:
merge main

ryantxu committed:
merge main

torkelo commented:
I have tested this quite a bit, with dashboard imports as well. looks good so far

need to check the alerting bits.

ryantxu committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

ryantxu commented on packages/grafana-data/src/types/query.ts:
```diff
@@ -0,0 +1,42 @@
+export enum DataTopic {
```
There are still LOTS that need cue schema -- and importantly lots where we need to understand if our typescript is what it should be.

This is a case where our current structure allows writing values we do not intend to be in the saved json

torkelo replied:
this enum is only used in the return data right now

torkelo replied:
not in any persisted model

ryantxu commented:
This is too risky to squeeze into 8.0 and we can apply this migration in 8.1+ ideally along with some cue model for datasources!

joshhunt commented on packages/grafana-data/src/types/dashboard.ts:
```diff
@@ -1,5 +1,34 @@
   Tooltip,
 }
+
+export interface PanelModel<TOptions = any> {
```
what do you think about using `unknown` instead to avoid developers making assumptions about untyped fields?

ryantxu replied:
I think it is worth trying -- I just plopped in `unknown` and typescript complains about some of the migrations.  Maybe we can try that in another PR.

ryantxu committed:
merge main

ryantxu committed:
merge main

ryantxu committed:
merge main

ryantxu committed:
compile and run

ryantxu committed:
no changes to docs

ryantxu committed:
no changes to docs

ryantxu committed:
merge main

ryantxu committed:
add ModelVersion

ryantxu commented on packages/grafana-data/src/types/query.ts:
```diff
@@ -8,11 +10,18 @@ export enum DataTopic {
+  uid?: string;
+
+  /** The schema version for query target JSON  */
+  modelVersion?: ModelVersion;
```
we are not using this yet, but it is important think though where this should be saved

ryantxu committed:
maybe

ryantxu committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

ryantxu committed:
a few more tests

ryantxu committed:
a few more tests

ryantxu committed:
more types

ryantxu committed:
another test

eltociear committed:
Annotations: Fix typo in types.ts (#37910)

ryantxu commented on public/app/features/alerting/unified/utils/rule-form.ts:
```diff
@@ -203,13 +207,13 @@ const dataQueriesToGrafanaQueries = async (
   relativeTimeRange: RelativeTimeRange,
   scopedVars: ScopedVars | {},
-  datasourceName?: string,
+  datasourceRef?: DatasourceRef,
```
hymm -- probably do not want to change this

ivanahuckova committed:
Add pattern parser to syntax (#37824)

stephaniehingtgen committed:
Update http error code when datasource health check fails (#37895)

Since the datasource settings are configured by the user, the health check error could be a client error. A 506 error may be sufficient as well.

domasx2 committed:
remove guideliens form rule list state table (#37839)

gabor committed:
Prometheus/Alerting: handle datasource.timeRange is nil case (#37888)

sherwind committed:
CloudWatch: Add AWS/EC2 MetadataNoToken metric (#37427)

dsotirakis committed:
Elasticsearch: Fix metric names for alert queries (#37871)

* Use props names as metrics names

* Make aliases work

olbo98 committed:
Loki: add support for resolution (#36710)

* Add input to specify min step

* Add stepInterval to as input to component

* Add onBlur to Input component

* Loki: add functionality for min step

* Loki: change name on props to step to make it more clear

* Loki: add resolution as a query option

* Loki: Add min,max,exact as step options

* Loki: add functionality for different step modes

* Loki: fix bug where step function isn't working

* Loki: fix bug where exact step isn't working

* Loki: change width of step input field

* Loki: add tests for adjustInterval function

* Loki: add check for max step oprio to make sure it's not below the safe interval

* Loki: fix bug with some tests

* Loki: fix bug with tests

* Explore: add tooltip to loki step function

* Loki: remove resolution as a logs option

* Loki: update snapshots

* Fix failing tests

* Loki: add select component for choosing resolution

* Loki: add functionality for calculating correct interval with resolution applied

* Loki: remove functionality for step mode

* Loki: remove tests for step mode

* Loki: add tooltip to line limit and resolution

* Loki: add backend support for resolution

* Loki: fixed backend bug where resolution was undefined

* Loki: add check for resolution

dsotirakis committed:
Chore: Update channel to where failed Docker nightly builds are posted  (#37927)

* Update channel

* Sign drone.yml

axelavargas committed:
Navigation: Add aria-label to top section links to improve a11y for screen readers (#36953)

tskarhed committed:
Search: Make search icon keyboard navigable (#37865)

* Make search icon keyboard navigable

* Update text

* Update test

* Remove unused line

* Add global focus-visible styles for button

ashharrison90 committed:
PanelEdit: Fix 'Actual' size by passing the correct panel size to DashboardPanel (#37885)

razor-x committed:
Add missing AWS/Cognito metrics (#37768)

* Add missing AWS/Cognito metrics

https://docs.aws.amazon.com/cognito/latest/developerguide/metrics-for-cognito-user-pools.html

* Add missing Cognito dimensions

* Remove Cognito usage metrics

atifali committed:
Live: No connection warning for user with no org (#37943)

ryantxu committed:
query type

ryantxu committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

ryantxu committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

ryantxu committed:
merge main

ryantxu committed:
merge main

ryantxu committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

ryantxu committed:
tried to merge

torkelo committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

torkelo committed:
fixing unified alerting type errors

torkelo committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

torkelo committed:
Fixing explore type issue

torkelo committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

torkelo committed:
Fixing alerting tests

torkelo committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

torkelo committed:
Fixing rule editor conditions lookup issue

torkelo committed:
Fixing adhoc variable issue

torkelo committed:
Fixing ad hoc filters issue

torkelo commented on public/app/features/templating/template_srv.ts:
```diff
@@ -92,14 +92,21 @@ export class TemplateSrv implements BaseTemplateSrv {
+  getAdhocFilters(datasourceName: string): AdHocVariableFilter[] {
     let filters: any = [];
+    let ds = getDataSourceSrv().getInstanceSettings(datasourceName);
+
+    if (!ds) {
+      return [];
+    }
 
     for (const variable of this.getAdHocVariables()) {
-      if (variable.datasource === null || variable.datasource === datasourceName) {
+      const variableUid = variable.datasource?.uid;
+
+      if (variableUid === ds.uid || (variable.datasource == null && ds?.isDefault)) {
         filters = filters.concat(variable.filters);
-      } else if (variable.datasource.indexOf('$') === 0) {
-        if (this.replace(variable.datasource) === datasourceName) {
+      } else if (variableUid?.indexOf('$') === 0) {
+        if (this.replace(variableUid) === datasourceName) {
           filters = filters.concat(variable.filters);
         }
       }
```
This function was a bit tricky, as it's part something many data sources use (internal & external) to support ad hoc queries. I opted to just lookup the data source by name to get the uid to find the matching adhoc variable. 

We could also change this signature to `getAdhocFilters(dsUid: string)` as long as we always lookup the ds via getInstanceSettings incase external plugins call this function with the data source name.

torkelo committed:
Changed name of DatasourceRef to DataSourceRef for consistency with DataSourceSettings, DataSourceAPI and DataSourcePlugin etc

torkelo committed:
Fix ts issue

torkelo commented:
Todo: 
New unified alerting rules fail here: https://github.com/grafana/grafana/blob/main/pkg/expr/graph.go#L148

torkelo committed:
Got new unified alerting working & service side expressions

torkelo committed:
Updates

torkelo committed:
fixing template_srv tests

torkelo committed:
fixing ts issue

torkelo committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

torkelo commented:
Datasource variables (variables that currently contain data source NAMES) are going to be a problem I think as they currently evaluate to the name of the data source no the UID.

torkelo committed:
Fixed migrator test

torkelo committed:
Fixing explore QueryRows test

torkelo committed:
fixing old alerting

torkelo committed:
Fixing explore test

torkelo committed:
Fixing DashboardExporter tests

torkelo commented:
* Fixed old alerting (updated the alert rule extractor & added new test)
* Fixed explore tests
* Fixed DashboardExport tests

Issues
* Dashboard export & import are not fully working, I mean they are but they sort of produce old schema, the dashboard exporter will overwrite datasource property and turn it back into a string. So the importer actually works as before, but you end up with a dashboard model (after import) that has schemaVersion of 32 but with old schema (datasource string properties). 

Not done
* annotation queries, but checked it looks like we can do this in a separate PR.

torkelo committed:
fix go test

torkelo committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

torkelo committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

torkelo committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

torkelo committed:
fixed test

marefr reviewed (commented):
Backend changes looks good based on a quick look

torkelo committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

torkelo reviewed (approved):
@ryantxu @sdboyer think this is look in good shape. One question I have is if DataSourceRef object should have a name property as well, for reference? To aid in as code scenarios and when reading dashboard  json models. Only having uid will make it hard for humans to reason about the model (understand what datasource is being used)

ryantxu committed:
merge main

sdboyer reviewed (commented):
Seems like we can probably defer my default-related concerns to a follow-up PR?

sdboyer reviewed (dismissed):
Actually gonna request change specifically about disallowing the null. We can defer that to later if it ends up inducing a bunch of extra work here, but i'd prefer we at least try to not allow the optional + nullable redundancy.

ryantxu committed:
use datasource ref

ryantxu committed:
use datasource ref

ryantxu committed:
change expected order

ryantxu committed:
revert order change

ryantxu committed:
with ref

mckn reviewed (commented):
Awesome work on this! Really nice to get better types in place. I left a couple of nits to consider.

ryantxu committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

ryantxu committed:
more cleanup

ryantxu committed:
fix mixed test

ryantxu committed:
use ref

ryantxu committed:
update imports

ryantxu committed:
update some comments

torkelo committed:
Fixed alerting test

torkelo committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

ryantxu committed:
fix mixed test

ryantxu committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

ryantxu committed:
getRef

ryantxu committed:
merge main

ryantxu committed:
Merge remote-tracking branch 'origin/main' into with-datasource-ref

ryantxu committed:
explore fix

Elfo404 committed:
Merge branch 'main' into with-datasource-ref

Elfo404 committed:
fix explore tests

torkelo reviewed (approved):
Let's do it :see_no_evil:

hugohaggmark commented:
@ryantxu the description of this PR needs to include the breaking change, I don't remember the exact format but I think you need to add something in the end of the description. Thank you 🙏

malcolmholmes commented:
I assume that this introduces a new schema version, and that older schema versions will be migrated to this version. In which case, at least in this sense, this *isn't* a breaking change. Existing dashboard json will continue to work. It will only break if you update the schema version without updating your dashboard references.

Are these assumptions correct?

hugohaggmark commented on public/app/features/dashboard/state/DashboardMigrator.ts:
```diff
@@ -62,7 +65,7 @@ export class DashboardMigrator {
     const oldVersion = this.dashboard.schemaVersion;
     const panelUpgrades: PanelSchemeUpgradeHandler[] = [];
-    this.dashboard.schemaVersion = 32;
+    this.dashboard.schemaVersion = 33;
```
@malcolmholmes this changes the schemaVersion

dginhoux commented:
Very great idea, thank you very mich for this improvment

marefr commented:
> I assume that this introduces a new schema version, and that older schema versions will be migrated to this version. In which case, at least in this sense, this _isn't_ a breaking change. Existing dashboard json will continue to work. It will only break if you update the schema version without updating your dashboard references.
> 
> Are these assumptions correct?

I removed the breaking change label for now.

jgarciadelanoceda commented:
I think that this is still a Breaking Change because we use Grafana in several environments and now it is not possible to copy a Dashboard to another environment.

If now you copy a dashboard to another environment now you need to change all the panels (because the uid is not always the same)
